### PR TITLE
Add automatic MAC address discovery for GAN iCarry2

### PR DESCRIPTION
**Pull request by guillaume86**
Hi, 

I used this sample project: https://gan-mac-poc.stackblitz.io/ and tinkered a bit with it to make the MAC address discovery work for my GAN iCarry 2 (actually just brute force requested all ids to 0xFFFF to find it).

Thanks for your project ;)

### CODE REVIEW

**Comment by afedotov** [comment]
Thats interesting...
My iCarry2 instance OOB has CIC = 0x0001, but after updating firmware to v5.5 CIC become equal to 0x1001. I suppose we can encounter a problem with lot more random CIC values appear on iCarry2 cubes. And this really can be a problem since with Web Bluetooth you must to specify exact device CIC in requestDevice call to able to read Manufacturer Specific Data, this is security limitation.

What firmware version your cube have? Did you update cube firmware?

BTW you don't need to bruteforce CIC. To get CIC of your cube device just use nRF Connect application on Android, or simply go to Chrome bluetooth debug page - `chrome://bluetooth-internals/#devices` and look into the `Manufacturer Data` table field.

**Comment by guillaume86** [comment]
Just checked, I'm on firmware 5.5, thanks for the tip for inspecting the device.
Web Bluetooth requires the CIC but it seems like it doesn't prevent requesting a lot of them at once, I discovered mine like this actually:
```typescript
//const BT_GAN_MANUFACTURER_ID_LIST: Array<number> = [0x0001, 0x1001, 0x0501, 0x7f01];
const BT_GAN_MANUFACTURER_ID_LIST = Array.from(new Array(0xffff)).map((x, i) => i);
```
Maybe you can use this as a fallback if normal discovery fails.

**Comment by afedotov** [comment]
LOL just checked my iCarry2, and its CIC back to 0x0001.
I suppose its value got reset after cube deep sleep / reboot.
Looks like one more firmware bug, looks like uninitialized memory/variable, so I suppose occasionally it may advertise random CIC.

**Comment by guillaume86** [comment]
Haha ok it's that kind of firmware :)

**Comment by afedotov** [comment]
> ```ts
> Array.from(new Array(0xffff)).map((x, i) => i);
> ```

Not sure how to safe is to pass such big array with 65k number objects to requestDevice(). It is better to avoid such pattern, and this should be carefully tested on all bluetooth-enabled browsers.

**Comment by guillaume86** [comment]
Yeah I understand it's a messy way to do it. I just tried finding the CIC in the chrome bluetooth view but can't seem to find it:
image
It should have been there?

**Comment by afedotov** [comment]
> It should have been there?

Yes, your cube should send advertising packets and Chrome must be in scanning mode. Only then it will display Manufacturer Data. If you stop scan that field will be cleared.

**Comment by guillaume86** [comment]
I can't seem to get anything to show up there from the chrome tools, restarted scanning several times, disconnected, forget, etc.

I checked again, the CIC didn't change yet for me, still 0x7f01. The name of my device is "GANicV2S_42D2", does it match yours?

**Comment by afedotov** [comment]
idk what are you doing wrong, just start scan for devices, do not connect, and look into device list.

**Comment by guillaume86** [comment]
Ha OK in this table it's actually flashing briefly, had to attach a MutationObserver to the table cell, to capture the value and got it: 0x7f01 0x000000d24249b520d0 so yeah it's possible to get it from there.

**Comment by afedotov** [comment]
> I checked again, the CIC didn't change yet for me, still 0x7f01. The name of my device is "GANicV2S_42D2", does it match yours?

I suppose cube must come to deep sleep or reboot. So just put it aside for a couple of hours without shake or movement (since it it awakened by gravity sensor), and check it later. I suppose eventually CIC may reset back as in my cube did.

Anyway if this is really a problem with lot of randomly appeared CICs, we need to think how to fix this. Maybe this CIC list can be shortened to just 255 items since only first CIC byte is changing.

**Comment by afedotov** [comment]
Regarding of using generated CIC list that contains all possible CIC values.
https://issues.chromium.org/issues/40718843
Potentially big CIC list may cause performance issues at least in Chrome. Although this marked as "Fixed" recently, it may take time until it will spread among running browsers in the wild.

**Comment by afedotov** [comment]
I figured out mystery behind different CIC values of the GAN cubes.

**TLDR:** Just unbind you cube from account by pressing button in the CubeStation, or just quickly rotate Orange face of the cube for 5 rounds (20 quarter turns). After this cube will back to advertising with default `0x0001` CIC.

This is how typical Manufacturer Specific Data field in the bluetooth advertising packet should looks like:

`0C FF CC CC DD DD DD DD DD DD DD DD DD`

* `0C` - field length
* `FF` - field type (0xFF == manufacturer specific data)
* `CC CC` - 16-bit LE Int with fixed Company Identifier Code
* `DD DD ..` - arbitrary manufacturer specific data bytes

But GAN doesn't give a fuck and use it in their cubes in the following way:

`0C FF 01 AA AA AA AA DD DD DD DD DD DD`

* `01` - just a fixed number == 0x01 for all Gen2 and Gen3 protocol cubes
* `AA ..` - 32-bit BE Int with user account id
* `DD ..` - 6 bytes of bluetooth MAC address

So MSB byte of CIC is used for MSB byte of account id number. When the cube is bound to the user account in the CubeStation, account id is saved to the cube and being sent in the advertising packet, otherwise there are just zero bytes. Since most (or even all) of the user account numbers are not such big, MSB byte is typically just equals to 0, so Gen2 cubes just advertise with same CIC = `0x0001` even being bound to the account.

But iCarry2 firmware has a lot of bugs. And even CubeStation too, it handle iCarry2 protocol in wrong way (what a shame) in some places using incorrectly formatted commands to the cube. In this case it results that account id is just incorrectly being written to the cube taking only LSB byte of the id and placing it to the MSB byte of the CIC. So if iCarry2 instance is bound to the account in the CubeStation, it will start advertise with some random CIC in range of `[0x0001, 0xFF01]`, where only MSB byte is changing depending on bound user account id LSB byte.

So we have two options:
* Keep CIC list as is, actually only `0x0001` is needed. And guide users to unbind the cube from account to be able to automatically read cube MAC address upon connection.
* Fill CIC list with 256 number values from mentioned range. But this should be tested on mainstream bluetooth-enabled browsers (at least on Chrome and Bluefy) to avoid performance issues or crashes.
